### PR TITLE
ipq40xx: Add support for Teltonika RUTX10

### PR DESCRIPTION
**Pull request by 0xFelix**
Blocked by https://github.com/openwrt/openwrt/pull/4335

This patch adds support for the Teltonika RUTX10.
This device is an industrial DIN-rail router with 4 ethernet ports,
2.4G/5G dualband WiFi, Bluetooth, a USB 2.0 port and two GPIOs.

The RUTX series devices are very similiar so common parts of the DTS
are kept in a DTSI file. They are based on the QCA AP-DK01.1-C1 dev
board.

See https://teltonika-networks.com/product/rutx10 for more info.

Hardware:
  SoC:                           Qualcomm IPQ4018
  RAM:                          256MB DDR3
  SPI Flash 1:               XTX XT25F128B (16MB, NOR)
  SPI Flash 2:               XTX XT26G02AWS (256MB, NAND)
  Ethernet:                    Built-in IPQ4018 (SoC, QCA8075), 4x 10/100/1000 ports
  WiFi 1:                       Qualcomm QCA4019 IEEE 802.11b/g/n
  Wifi 2:                        Qualcomm QCA4019 IEEE 802.11a/n/ac
  USB Hub:                  Genesys Logic GL852GT
  Bluetooth:                  Qualcomm CSR8510 (A10U)
  LED/GPIO controller: STM32F030 with custom firmware
  Buttons:                     Reset button
  Leds:                         Power (green, cannot be controlled)
                                    WiFi 2.4G activity (green)
                                    WiFi 5G activity (green)

MACs Details verified with the stock firmware:
   eth0:                     Partition 0:CONFIG Offset: 0x0
   eth1:                     = eth0 + 1
   radio0 (2.4 GHz): = eth0 + 2
   radio1 (5.0 GHz): = eth0 + 3
Label MAC address is from eth0.

The LED/GPIO controller needs a separate kernel driver to function.
The driver was extracted from the Teltonika GPL sources and can be
found at following feed: https://github.com/0xFelix/teltonika-rutx-openwrt

USB detection of the bluetooth interface is sometimes a bit flaky. When
not detected power cycle the device. When the bluetooth interface was
detected properly it can be used with bluez / bluetoothctl.

Flash instructions via stock web interface (sysupgrade based):
  1. Set PC to fixed ip address 192.168.1.100
  2. Push reset button and power on the device
  3. Open u-boot HTTP recovery at http://192.168.1.1
  4. Upload latest stock firmware and wait until the device is rebooted
  5. Open stock web interface at http://192.168.1.1
  6. Set some password so the web interface is happy
  7. Go to firmware upgrade settings
  8. Choose
     openwrt-ipq40xx-generic-teltonika_rutx10-squashfs-nand-factory.bin
  9. Set 'Keep settings' to off
  10. Click update, when warned that it is not a signed image proceed

Return to stock firmware:
  1. Set PC to fixed ip address 192.168.1.100
  2. Push reset button and power on the device
  3. Open u-boot HTTP recovery at http://192.168.1.1
  4. Upload latest stock firmware and wait until the device is rebooted

Note: The DTS expects OpenWrt to be running from the second rootfs partition.
u-boot on these devices hot-patches the DTS so running from the first rootfs
partition should also be possible. If you want to be save follow the instructions
above. u-boot HTTP recovery restores the device so that when flashing OpenWrt
from stock firmware it is flashed to the second rootfs partition and the DTS matches.

### CODE REVIEW

File: target/linux/ipq40xx/base-files/etc/board.d/02_network [hunk 1]
```diff
@@ -181,6 +182,11 @@ ipq40xx_setup_macs()
 		lan_mac=$(cat /sys/firmware/mikrotik/hard_config/mac_base)
 		label_mac="$lan_mac"
 		;;
+	teltonika,rutx10)
+		lan_mac=$(mtd_get_mac_binary "0:CONFIG" 0x0)
```
Comment (adschm) [inline target/linux/ipq40xx/base-files/etc/board.d/02_network:186]: Can't this be set via DTS?

File: target/linux/ipq40xx/base-files/lib/upgrade/platform.sh [hunk 2]
```diff
@@ -124,6 +124,10 @@ platform_do_upgrade() {
 		PART_NAME="inactive"
 		platform_do_upgrade_dualboot_datachk "$1"
 		;;
+  teltonika,rutx10)
```
Comment (adschm) [inline target/linux/ipq40xx/base-files/lib/upgrade/platform.sh:127]: Broken indent.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx.dtsi [hunk 3]
```diff
@@ -0,0 +1,228 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "qcom-ipq4019-ap.dk01.1.dtsi"
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx.dtsi:6]: This should probably be moved before the other includes, so the dts-v1 markup is first.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx10.dts [hunk 4]
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qcom-ipq4018-rutx.dtsi"
+#include <dt-bindings/gpio/gpio.h>
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx10.dts:4]: gpio.h is already in dtsi?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx10.dts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qcom-ipq4018-rutx.dtsi"
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx10.dts:3]: Why do we need a DTSI at all here, just for one device?

File: target/linux/ipq40xx/image/generic.mk [hunk 6]
```diff
@@ -780,6 +780,25 @@ define Device/qxwlan_e2600ac-c2
...
+	DEVICE_VENDOR := Teltonika
+	DEVICE_MODEL := RUTX10
+	SOC := qcom-ipq4018
+	DEVICE_DTS_CONFIG := config@5
+	KERNEL_INSTALL := 1
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	FILESYSTEMS := squashfs
+	SUPPORTED_DEVICES := teltonika,rutx10 teltonika,rutx
+	IMAGE_METADATA := "device_code": [".*"], "hwver": [".*"], "batch": [".*"], "serial": [".*"]
```
Comment (adschm) [inline target/linux/ipq40xx/image/generic.mk:795]: What's that supposed to do? I've never seen anything like it?

File: target/linux/ipq40xx/image/generic.mk [hunk 7]
```diff
@@ -780,6 +780,25 @@ define Device/qxwlan_e2600ac-c2
...
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := Teltonika
+	DEVICE_MODEL := RUTX10
+	SOC := qcom-ipq4018
+	DEVICE_DTS_CONFIG := config@5
+	KERNEL_INSTALL := 1
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	FILESYSTEMS := squashfs
+	SUPPORTED_DEVICES := teltonika,rutx10 teltonika,rutx
```
Comment (adschm) [inline target/linux/ipq40xx/image/generic.mk:794]: First string is default, what's the second needed for?
Probably this line can be removed completely ...

**Review by adschm — CHANGES_REQUESTED** [review]

File: target/linux/ipq40xx/base-files/etc/board.d/02_network (hunk 1 above)
Comment (0xFelix) [reply target/linux/ipq40xx/base-files/etc/board.d/02_network]: Yes! Changed it

File: target/linux/ipq40xx/base-files/lib/upgrade/platform.sh (hunk 2 above)
Comment (0xFelix) [reply target/linux/ipq40xx/base-files/lib/upgrade/platform.sh]: Fixed

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx.dtsi (hunk 3 above)
Comment (0xFelix) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx.dtsi]: Fixed

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx10.dts (hunk 4 above)
Comment (0xFelix) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx10.dts]: Fixed

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx10.dts (hunk 5 above)
Comment (0xFelix) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx10.dts:3]: Teltonika has several RUTX devices which share the same platform. I already separated the common parts into the DTSI. Makes adding other devices easier.

File: target/linux/ipq40xx/image/generic.mk (hunk 6 above)
Comment (0xFelix) [reply target/linux/ipq40xx/image/generic.mk:795]: This is needed for the stock firmware to accept the factory image. Stock firmware is QSDK based and uses a modified sysupgrade. It checks the metadata for these JSON nodes.

File: target/linux/ipq40xx/image/generic.mk (hunk 7 above)
Comment (0xFelix) [reply target/linux/ipq40xx/image/generic.mk:794]: This is needed for the stock firmware to accept the factory image. Stock firmware checks for `teltonika,rutx` in metadata.

File: target/linux/ipq40xx/image/generic.mk (hunk 6 above)
Comment (adschm) [reply target/linux/ipq40xx/image/generic.mk:795]: Please don't hack the OpenWrt metadata for factory flashing. Add a proper factory image with its own recipe instead.

File: target/linux/ipq40xx/image/generic.mk (hunk 7 above)
Comment (adschm) [reply target/linux/ipq40xx/image/generic.mk:794]: If possible, this should be handled by a factory image instead.

File: target/linux/ipq40xx/image/generic.mk (hunk 6 above)
Comment (0xFelix) [reply target/linux/ipq40xx/image/generic.mk]: I added an `append-rutx-metadata`  image command

File: target/linux/ipq40xx/image/generic.mk (hunk 7 above)
Comment (0xFelix) [reply target/linux/ipq40xx/image/generic.mk]: This is handled by the factory image now

**Comment by 0xFelix** [comment]
@adschm Adressed your change requests, looks good now?

**Comment by 0xFelix** [comment]
https://github.com/openwrt/openwrt/pull/4335 was merged, this PR is now no longer blocked

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 8]
```diff
@@ -154,6 +154,10 @@ case "$FIRMWARE" in
 		caldata_extract "0:ART" 0x1000 0x2f20
 		ath10k_patch_mac $(mtd_get_mac_binary "0:MANUDATA" 0x6)
 		;;
+	teltonika,rutx10)
+		caldata_extract "0:ART" 0x1000 0x2f20
+		ath10k_patch_mac $(macaddr_add $(get_mac_label) 2)
```
Comment (adschm) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:159]: Please don't use get_mac_label here, it will result in undefined behavior.

Either read from flash or from eth0 if not possible otherwise.

File: target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 9]
```diff
@@ -268,6 +272,10 @@ case "$FIRMWARE" in
 		caldata_extract "0:ART" 0x5000 0x2f20
 		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary "0:MANUDATA" 0x6) 16)
 		;;
+	teltonika,rutx10)
+		caldata_extract "0:ART" 0x5000 0x2f20
+		ath10k_patch_mac $(macaddr_add $(get_mac_label) 3)
```
Comment (adschm) [inline target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:277]: See above.

File: target/linux/ipq40xx/image/generic.mk [hunk 10]
```diff
@@ -780,6 +791,23 @@ define Device/qxwlan_e2600ac-c2
...
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := Teltonika
+	DEVICE_MODEL := RUTX10
+	SOC := qcom-ipq4018
+	DEVICE_DTS_CONFIG := config@5
+	KERNEL_INSTALL := 1
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	FILESYSTEMS := squashfs
+	IMAGES := nand-factory.bin nand-sysupgrade.bin
```
Comment (adschm) [inline target/linux/ipq40xx/image/generic.mk:805]: This line can be dropped, it's already set it UbiFit.

**Comment by 0xFelix** [comment]
Changed as requested

**Comment by 0xFelix** [comment]
@adschm Ping :)

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx.dtsi [hunk 11]
```diff
@@ -0,0 +1,251 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qcom-ipq4019-ap.dk01.1.dtsi"
```
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx.dtsi:3]: Are these devices actually related?

Comment (adschm) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-rutx.dtsi:3]: Never mind, just read the commit message ...

**Comment by 0xFelix** [comment]
@adschm Does the PR still need changes?

**Comment by hauke** [comment]
Thank you for the patch, I applied it to master in 1cc3b95efcc616fba91a7f34e9781743229cc640
